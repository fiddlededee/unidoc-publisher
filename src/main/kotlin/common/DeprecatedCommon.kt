package common

@Deprecated("Renamed to converter.fodt.AsciidoctorOdAdapter")
typealias AsciidoctorAdapter = converter.fodt.AsciidoctorOdAdapter

