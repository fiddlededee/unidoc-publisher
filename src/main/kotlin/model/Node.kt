package model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator

fun document(init: Document.() -> Unit = {}): Document {
    return Document().apply(init)
}

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonTypeInfo(include = JsonTypeInfo.As.WRAPPER_OBJECT, use = JsonTypeInfo.Id.NAME)
abstract class Node() {

    var id: String? = null
    var sourceTagName: String? = null
    var includeTags: MutableSet<String> = mutableSetOf()
    var sourceAttributes: MutableMap<String, String> = mutableMapOf()
    var sourceMapping: SourceMapping? = null

    @get:JsonIgnore
    abstract val isInline: Boolean

    @JsonIgnore
    private var parent: Node? = null

    private val children: ArrayList<Node> = arrayListOf()
    val roles = arrayListOf<String>()
    fun <T : Node> appendChild(childNode: T): T {
        val oldParent = childNode.parent
        children.add(childNode)
        childNode.parent = this
        oldParent?.removeChild(childNode)
        return childNode
    }

    fun <T : Node> appendFirstChild(childNode: T): T {
        if (children().size == 0)
            appendChild(childNode) else
            children()[0].insertBefore(childNode)
        return childNode
    }

    fun table(init: Table.() -> Unit = {}): Table {
        return appendChild(Table().apply(init))
    }

    fun img(src: String, init: Image.() -> Unit = {}): Image {
        return appendChild(Image(src).apply(init))
    }

    fun tableRowGroup(type: TRG, init: TableRowGroup.() -> Unit = {}): TableRowGroup {
        return appendChild(TableRowGroup(type).apply(init))
    }

    fun col(width: Length): Col {
        return appendChild(Col(width))
    }

    fun tr(init: TableRow.() -> Unit = {}): TableRow {
        return appendChild(TableRow().apply(init))
    }

    fun td(init: TableCell.() -> Unit = {}): TableCell {
        return appendChild(TableCell().apply(init))
    }

    fun h(level: Int, init: Heading.() -> Unit = {}): Heading {
        return appendChild(Heading(level).apply(init))
    }


    fun li(init: ListItem.() -> Unit = {}): ListItem {
        return appendChild(ListItem().apply(init))
    }

    fun ol(init: OrderedList.() -> Unit = {}): OrderedList {
        return appendChild(OrderedList().apply(init))
    }

    fun ul(init: UnorderedList.() -> Unit = {}): UnorderedList {
        return appendChild(UnorderedList().apply(init))
    }

    fun p(init: Paragraph.() -> Unit = {}): Paragraph {
        return appendChild(Paragraph().apply(init))
    }

    fun span(init: Span.() -> Unit = {}): Span {
        return appendChild(Span().apply(init))
    }

    fun a(href: String, init: Anchor.() -> Unit = {}): Anchor {
        return appendChild(Anchor(href).apply(init))
    }

    fun text(text: String): Text {
        return appendChild(Text(text))
    }

    fun openBlock(init: OpenBlock.() -> Unit = {}): OpenBlock {
        return appendChild(OpenBlock().apply(init))
    }

    operator fun String.unaryPlus(): Text {
        return this@Node.appendChild(Text(this))
    }

    private fun removeChild(childNode: Node) {
        children.remove(childNode)
    }

    fun remove() {
        val parentNode = parent
        if (parentNode != null) {
            parentNode.removeChild(this)
        } else throw Exception("Node doesn't hava a parent")
    }

    fun insertAfter(nodeToInsert: Node): Node {
        val oldParent = nodeToInsert.parent
        val parent = this.parent
            ?: throw Exception("Can't add node after another if it has no parent")
        val index = parent.children.indexOf(this)
        if (index == -1) {
            throw Exception("Didn't find object to add after")
        }
        parent.children.add(index + 1, nodeToInsert)
        nodeToInsert.parent = parent
        oldParent?.removeChild(nodeToInsert)
        return nodeToInsert
    }

    fun insertBefore(nodeToInsert: Node): Node {
        val oldParent = nodeToInsert.parent
        val parent = this.parent
            ?: throw Exception("Can't add node before another if it has no parent")
        val index = parent.children.indexOf(this)
        if (index == -1) {
            throw Exception("Didn't find object to add after")
        }
        parent.children.add(index, nodeToInsert)
        nodeToInsert.parent = parent
        oldParent?.removeChild(nodeToInsert)
        return nodeToInsert
    }


    // todo abstract ?
    abstract fun write(bw: BackendWriter)

    private fun ArrayList<Node>.fillWithDescendants(node: Node, filter: (Node) -> Boolean) {
        node.children.forEach {
            if (filter(it)) {
                add(it)
            }
            fillWithDescendants(it, filter)
        }
    }

    fun descendant(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        return arrayListOf<Node>().apply { fillWithDescendants(this@Node, filter) }
    }

    fun ancestor(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val ancestors = arrayListOf<Node>()
        var parent = this@Node.parent()
        while (parent != null) {
            if (filter.invoke(parent)) ancestors.add(parent)
            parent = parent.parent()
        }
        return ancestors
    }

    fun nextSibling(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val nextSibling = arrayListOf<Node>()
        val parent = this.parent() ?: return nextSibling
        if (parent.children.indexOf(this) == parent.children.size - 1) return nextSibling
        IntRange(parent.children.indexOf(this) + 1, parent.children.size - 1).forEach {
            if (filter(parent.children[it])) nextSibling.add(parent.children[it])
        }
        return nextSibling
    }

    fun previousSibling(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val previousSibling = arrayListOf<Node>()
        val parent = this.parent() ?: return previousSibling
        if (parent.children.indexOf(this) == 0) return previousSibling
        IntRange(0, parent.children.indexOf(this) - 1).forEach {
            if (filter(parent.children[it])) previousSibling.add(parent.children[it])
        }
        return previousSibling
    }
    fun hasNext(): Boolean {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        return (parent.children.indexOf(this) != parent.children.size - 1)
    }

    fun next(): Node {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        if (parent.children.indexOf(this) == parent.children.size - 1)
            throw Exception("The ${this::class.java.simpleName} is last")
        return parent.children[parent.children.indexOf(this) + 1]
    }

    fun hasPrevious(): Boolean {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        return (parent.children.indexOf(this) != 0)
    }

    fun previous(): Node {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        if (parent.children.indexOf(this) == 0)
            throw Exception("The ${this::class.java.simpleName} is first")
        return parent.children[parent.children.indexOf(this) - 1]
    }

    fun index(): Int {
        val parent = this.parent()
            ?: throw Exception("The ${this::class.java.simpleName} node has no parent")
        return parent.children.indexOf(this)
    }

    @JsonProperty("children")
    fun getChildren(): ArrayList<Node> {
        return children
    }

    fun children(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        return arrayListOf<Node>()
            .apply { children.forEach { if (filter(it)) this.add(it) } }
    }

    fun parent(): Node? {
        return parent
    }

    fun roles(vararg role: String) {
        role.forEach { roles.add(it) }
    }

    fun toYamlString(): String {
        return ObjectMapper(YAMLFactory().disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER))
            .writeValueAsString(this)
    }

    fun <T : Node> T.wrapNodeInlineContents(): T {
        // Wrapping each inline nodes chains into paragraph node
        fun wrap(itemsToWrap: ArrayList<ArrayList<Node>>) {
            itemsToWrap.forEach { itemsChain ->
                if (itemsChain.isEmpty()) return@forEach
                itemsChain[0].insertBefore(Paragraph())
                    .apply { itemsChain.forEach { itemToWrap -> appendChild(itemToWrap) } }
            }
        }

        // Finding inline nodes chains
        val itemsToWrap = arrayListOf<ArrayList<Node>>()
        itemsToWrap.add(arrayListOf())
        this.children.forEach { nodeChild ->
            if (nodeChild.isInline) itemsToWrap.last().add(nodeChild)
            else if (itemsToWrap.last().isNotEmpty()) itemsToWrap.add(arrayListOf())
        }
        wrap(itemsToWrap)
        return this
    }


    // Don't understand html rendering rules, hope I'm right about this
    fun normalizeWhitespaces() {
        val localAst = this
        // Cleaning whitespaces after and before block nodes
        localAst.descendant { it is Text && it.text.isBlank() }.forEach { blankTextNode ->
            val localParent = blankTextNode.parent() ?: return@forEach
            val previousConsideredBlock =
                (!blankTextNode.hasPrevious() && !localParent.isInline) ||
                        (blankTextNode.hasPrevious() && (
                                !blankTextNode.previous().isInline ||
                                        blankTextNode.previous().sourceTagName == "br"
                                ))
            val nextConsideredBlock =
                (!blankTextNode.hasNext() && !localParent.isInline) ||
                        (blankTextNode.hasNext() && (
                                !blankTextNode.next().isInline ||
                                        blankTextNode.next().sourceTagName == "br"
                                ))

            if (previousConsideredBlock || nextConsideredBlock)
                blankTextNode.remove()
        }
        // Removing extra whitespaces
        Document().apply { appendChild(localAst) }.descendant { !it.isInline }.forEach { block ->
            if (block.sourceTagName == "pre" || block.ancestor { it.sourceTagName == "pre" }.isNotEmpty())
                return@forEach
            val nodesToNormalise = block.descendant { node ->
                (node is Image || node is Text || !node.isInline) &&
                        node.ancestor { !it.isInline }.firstOrNull() == block
            }
            var chain: ArrayList<Node> = arrayListOf()
            nodesToNormalise.forEach searchForChainsToNormalize@{ node ->
                if (!node.isInline) {
                    if (chain.isEmpty()) return@searchForChainsToNormalize else {
                        normalizeChain(chain)
                        chain = arrayListOf()
                    }
                } else chain.add(node)
            }
            normalizeChain(chain)
        }
    }

    private fun normalizeChain(chain: ArrayList<Node>) {
        if (chain.isEmpty()) return
        var trimStart = true
        chain.forEach byNodesToNormalize@{ node ->
            if (node is Image || !node.isInline) trimStart = false
            if (node !is Text) return@byNodesToNormalize
            if (trimStart) {
                node.text = node.text.trimStart()
            } else {
                if (node.text.trimStart() != node.text) node.text = " " + node.text.trimStart()
            }
            if (node.text.trimEnd() != node.text && node.text.isNotEmpty()) {
                node.text = node.text.trimEnd() + " "
                trimStart = true
            } else if (node.text.isNotEmpty()) trimStart = false
        }
        val iterator = chain.reversed().iterator()
        while (iterator.hasNext()) {
            val el = iterator.next()
            if (el !is Text) break
            if (el.text.isBlank()) el.remove() else {
                el.text = el.text.trimEnd()
                break
            }
        }

    }

    fun extractText(replaceRule: (Text) -> String = { it.text }): String {
        if (this is Text) return replaceRule.invoke(this)
        return descendant { it is Text }
            .map { it as Text }
            .joinToString("") { text -> replaceRule.invoke(text) }
    }

}

