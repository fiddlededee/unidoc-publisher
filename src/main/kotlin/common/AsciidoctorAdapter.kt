package common

import model.*
import reader.HtmlNode
import reader.UnknownTagProcessing
import writer.*

object AsciidoctorAdapter : GenericAdapter {
    enum class Color { RED, GREEN, YELLOW, BLUE, ORANGE }

    var lightColors = arrayOf(
        Color.RED to "#f8ced0", Color.GREEN to "#c2f3d3",
        Color.YELLOW to "#fef0c2", Color.BLUE to "#c2f3f1",
        Color.ORANGE to "#fedaca"
    )
    var darkColors = arrayOf(
        Color.RED to "#8f141b", Color.GREEN to "#19803d",
        Color.YELLOW to "#a17c02", Color.BLUE to "#19807c",
        Color.ORANGE to "#a03203"
    )
    var admonitionColors = arrayOf(
        "note" to Color.BLUE,
        "tip" to Color.GREEN,
        "important" to Color.YELLOW,
        "caution" to Color.ORANGE,
        "warning" to Color.RED
    )

    enum class AdmonitionCellType { TYPE_CONTENT, CONTENT }

    override val unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing = {
        if (classNames().intersect(
                setOf(
                    "listingblock",
                    "admonitionblock",
                    "exampleblock",
                    "imageblock",
                    "paragraph",
                    "book",
                    "article",
                    "title",
                    "footnotes",
                    "footnote",
                    "dlist",
                    "olist",
                    "ulist",
                    "content",
                    *(1..6).map { "sect$it" }.toTypedArray(),
                )
            ).isNotEmpty()
        ) {
            UnknownTagProcessing.PROCESS
        } else if (setOf("dl", "dt", "dd").contains(nodeName())) {
            UnknownTagProcessing.PROCESS
        } else if (nodeName() == "div") {
            UnknownTagProcessing.PASS
        } else UnknownTagProcessing.UNDEFINDED
    }

    override fun basicStyle(): OdtStyleList {
        return OdtStyleList(
            OdtStyle { genericTitle ->
                if (genericTitle !is Paragraph) return@OdtStyle
                val genericTitleWrapper = genericTitle.parent()
                if (genericTitleWrapper == null ||
                    !genericTitleWrapper.roles.contains("title")
                ) return@OdtStyle
                attributes("text:style-name" to "Text")
            },
            OdtStyle { tableCell ->
                if (tableCell !is TableCell) return@OdtStyle
                tableCellProperties {
                    arrayOf("top", "right", "bottom", "left").forEach {
                        attribute("fo:border-$it", "0.5pt solid #000000")
                        if (it != "bottom") attribute("fo:padding-$it", "1mm")
                    }
                    if (tableCell.sourceTagName == "th") attribute("fo:background-color", "#eeeeee")
                }
                arrayOf("top", "middle", "bottom").forEach {
                    if (tableCell.roles.contains("valign-$it"))
                        tableCellProperties { attributes("style:vertical-align" to it) }
                }
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                if (paragraph.ancestor { it is TableCell }.isEmpty()) return@OdtStyle
                attributes("text:style-name" to "Table Contents")
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                arrayOf("center", "left", "right").forEach {
                    if (paragraph.parent()?.roles?.contains("halign-$it") == true)
                        paragraphProperties { attributes("fo:text-align" to it) }
                }
            },
            OdtStyle { tableCell ->
                if (tableCell !is TableCell) return@OdtStyle
                val nodeParent = tableCell.parent() ?: return@OdtStyle
                lightColors.forEach { color ->
                    if (nodeParent.descendant()
                            .map { it.roles }.reduce { acc, roles ->
                                ArrayList(acc + roles)
                            }.contains("row-background-color-${color.first.toString().lowercase()}")
                    ) {
                        tableCellProperties { attributes("fo:background-color" to color.second) }
                    }
                }

            },
            OdtStyle { table ->
                if (table !is Table) return@OdtStyle
                if (table.ancestor { it is Table }.isEmpty())
                    tableProperties { attributes("fo:margin-bottom" to "2mm") } else
                    tableProperties { attributes("fo:margin-bottom" to "1mm") }
            },
            OdtStyle { admonitionTypeContentCell ->
                if (admonitionTypeContentCell !is TableCell) return@OdtStyle
                val admonitionCellType = run {
                    if (admonitionTypeContentCell.roles.contains("admonition-type-content")) {
                        AdmonitionCellType.TYPE_CONTENT
                    } else if (admonitionTypeContentCell.roles.contains("admonition-content"))
                        AdmonitionCellType.CONTENT
                    else null
                }
                if (admonitionCellType == null) return@OdtStyle
                val admonitionBlockRoles = admonitionTypeContentCell
                    .ancestor { it.roles.contains("admonitionblock") }
                    .firstOrNull()
                    ?.roles ?: throw Exception("Not in an admonition block")
                admonitionColors.filter { admonitionBlockRoles.contains(it.first) }
                    .first()
                    .apply {
                        tableCellProperties {
                            val color =
                                lightColors.toMap()[second]
                                    ?: throw Exception("Color ${second} not found")
                            if (admonitionCellType == AdmonitionCellType.TYPE_CONTENT)
                                attributes("fo:background-color" to color)
                            arrayOf("left", "right", "top", "bottom").forEach {
                                attributes("fo:border-$it" to "0.7pt solid $color")
                            }
                        }
                    }
            },
            OdtStyle { admonitionTitleParagraph ->
                if (admonitionTitleParagraph !is Paragraph) return@OdtStyle
                if (admonitionTitleParagraph.ancestor { it.roles.contains("admonition-content") }.isEmpty())
                    return@OdtStyle
                if (admonitionTitleParagraph.ancestor { it.roles.contains("title") }.isEmpty())
                    return@OdtStyle
                textProperties { attributes("fo:font-style" to "italic") }
            },
            OdtStyle { titleParagraph ->
                if (titleParagraph !is Paragraph) return@OdtStyle
                if (!titleParagraph.roles.contains("doctitle")) return@OdtStyle
                attributes("text:style-name" to "Title")
            },
            OdtStyle { figureTitle ->
                if (figureTitle !is Paragraph) return@OdtStyle
                if (figureTitle.ancestor { it.roles.contains("imageblock") }.isEmpty()) return@OdtStyle
                if (figureTitle.ancestor { it.roles.contains("title") }.isEmpty()) return@OdtStyle
                attributes("text:style-name" to "Figure")
            },
            OdtStyle { imageblockParagraph ->
                if (imageblockParagraph !is Paragraph) return@OdtStyle
                if (!imageblockParagraph.roles.contains("imageblock-paragraph"))
                    return@OdtStyle
                attributes("text:style-name" to "Image Block")
                val imageBlockRoot =
                    imageblockParagraph.ancestor { it.roles.contains("imageblock") }
                        .firstOrNull()
                if (imageBlockRoot != null &&
                    imageBlockRoot.descendant { it.roles.contains("title") }.isNotEmpty()
                ) paragraphProperties { attributes("fo:keep-with-next" to "always") }
            },
            OdtStyle { tableTitle ->
                if (tableTitle !is Paragraph) return@OdtStyle
                val parent = tableTitle.parent()
                if (parent != null && parent.roles.contains("table-title")) {
                    attributes("text:style-name" to "Table")
                }
            },
            OdtStyle { listTitle ->
                if (listTitle !is Paragraph) return@OdtStyle
                val listTitleWrapper = listTitle.parent() ?: return@OdtStyle
                if (!listTitleWrapper.roles.contains("title")) return@OdtStyle
                if (!listTitleWrapper.hasNext()) return@OdtStyle
                val listTitleWrapperNextSibling = listTitleWrapper.next()
                if (listTitleWrapperNextSibling !is UnorderedList &&
                    listTitleWrapperNextSibling !is OrderedList
                ) return@OdtStyle
                if (listTitle.ancestor { it is Table }.isEmpty())
                    attributes("text:style-name" to "Text body") else
                    attributes("text:style-name" to "Table Contents")
                paragraphProperties { attributes("fo:keep-with-next" to "always") }
            },
            OdtStyle { listingTitle ->
                if (listingTitle !is Paragraph) return@OdtStyle
                val listingTitleWrapper = listingTitle.parent() ?: return@OdtStyle
                if (!listingTitleWrapper.roles.contains("title")) return@OdtStyle
                if (!listingTitleWrapper.hasNext()) return@OdtStyle
                val listingTitleWrapperNextSibling = listingTitleWrapper.next()
                if (!listingTitleWrapperNextSibling.roles.contains("pre")) return@OdtStyle
                attributes("text:style-name" to "Text")
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                if (!paragraph.roles.contains("definition-term")) return@OdtStyle
                attributes("text:style-name" to "Definition Term")
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                if (!paragraph.roles.contains("horizontal-line-for-example")) return@OdtStyle
                attributes("text:style-name" to "Horizontal Line for Example")
            },
            OdtStyle { node ->
                if (node.ancestor { it.roles.contains("keep-with-next") }.isNotEmpty()
                    || node.roles.contains("keep-with-next")
                ) {
                    if (node is Paragraph)
                        paragraphProperties { attributes("fo:keep-with-next" to "always") }
                    if (node is Table)
                        tableProperties { attributes("fo:keep-with-next" to "always") }
                    if (node is TextFrame)
                        graphicProperties { attributes("fo:keep-with-next" to "always") }
                }
            },
            OdtStyle { span ->
                if (span !is Span) return@OdtStyle
                darkColors.forEach {
                    if (span.roles.contains(it.first.toString().lowercase())) {
                        textProperties { attributes("fo:color" to it.second) }
                    }
                }
            },
            OdtStyle { paragraph ->
                if (paragraph !is Paragraph) return@OdtStyle
                if (!paragraph.roles.contains("text-frame-wrapper-paragraph")) return@OdtStyle
                textProperties { attributes("fo:font-size" to "0pt") }
                paragraphProperties { attributes("fo:font-size" to "0pt") }
            }
        )
    }

    fun Node.trimSimpleTdContent() {
        this.descendant { it is TableCell }.map { it as TableCell }
            .forEach { tableCell ->
                if (tableCell.children().size == 1 && tableCell.children()[0] is Text) {
                    val textNode = tableCell.children()[0] as Text
                    textNode.text = textNode.text.trim()
                }
            }
    }

    @JvmStatic
    fun Node.wrapTableCellInlineContent() {
        this.descendant { it is TableCell }.map { it as TableCell }
            .forEach { tableCell -> tableCell.wrapNodeInlineContents() }
    }

    @JvmStatic
    fun Node.wrapTitleInlineContent() {
        this.descendant { it is OpenBlock && it.roles.contains("title") }
            .forEach { openBlock ->
                openBlock.wrapNodeInlineContents()
                val table = openBlock.parent()
                if (table != null && table is Table) {
                    val tableId = table.id
                    if (tableId != null) {
                        openBlock.descendant { it is Paragraph }.firstOrNull()?.apply {
                            id = tableId
                        }
                        table.id = "table--${tableId}"
                    }
                    table.insertBefore(openBlock)
                    openBlock.roles("table-title")
                }
            }
    }

    @JvmStatic
    fun Node.extractToc() {
        val parsedToc = this.descendant { it.id == "toc" }
            .firstOrNull() ?: return
        val toc = Toc().apply {
            levels = parsedToc.descendant { it is UnorderedList }
                .maxOfOrNull { ul -> ul.ancestor { it is UnorderedList }.size + 1 }
                ?: throw Exception("Can't determine number of levels in the TOC")
            titleNode = Paragraph().apply {
                parsedToc.descendant { it.parent()?.id == "toctitle" }.forEach {
                    appendChild(it)
                }
            }
        }
        parsedToc.insertBefore(toc)
        parsedToc.remove()
    }


    fun Node.transformAdmonitions() {
        this.descendant { it.roles.contains("admonitionblock") }.forEach { block ->
            val oldTable = block.children().first { it is Table }
            val (typeContent, content) = run {
                oldTable
                    .children().first { it is TableRowGroup }
                    .children().first { it is TableRow }
                    .apply {
                        if (children().size != 2 ||
                            children().any { it !is TableCell }
                        ) throw Exception("Can't define admonition block")
                    }.children()
            }
            typeContent.roles("admonition-type-content")
            content.roles("admonition-content")
            block.apply {
                table {
                    col(Length(100F))
                    tableRowGroup(TRG.body) {
                        tr { appendChild(typeContent) }
                        tr { appendChild(content) }
                    }
                }
            }
            oldTable.remove()
        }
    }

    @JvmStatic
    fun Node.normalizeHeaders() {
        this.descendant { it is Header }.map { it as Header }.forEach { header ->
            if (header.level == 1) {
                val titleParagraph = Paragraph().apply {
                    roles("doctitle")
                    header.children().forEach { appendChild(it) }
                }
                header.insertBefore(titleParagraph)
                header.remove()
            } else header.level -= 1
        }
    }

    @JvmStatic
    fun Node.wrapTablesInLists() {
        this.descendant { it is ListItem }.forEach { listItem ->
            listItem.descendant().forEach liChildren@{ table ->
                if (table !is Table) return@liChildren
                if (table.parent() is TextFrame) return@liChildren
                val textFrame = TextFrame()
                val textFrameWrapperParagraph = Paragraph().apply {
                    roles("text-frame-wrapper-paragraph")
                    appendChild(textFrame)
                }
                table.insertBefore(textFrameWrapperParagraph)
                textFrame.appendChild(table)
            }
        }
    }


    @JvmStatic
    fun Node.extractFootnotes() {
        val footnotes: MutableMap<String, Footnote> = mutableMapOf()
        this.descendant { it is Span && it.roles.contains("footnote") }
            .forEach { footNoteSpan ->
                val footnoteId = footNoteSpan.id ?: throw Exception("Footnote id not found")
                val footnoteHref = footNoteSpan.descendant { it is Anchor }.map { it as Anchor }
                    .firstOrNull()
                    .apply { if (this == null) throw Exception("Anchor to footnote not found") }!!
                    .href
                    .substring(1)
                val footnoteBody = footNoteSpan.ancestor().last()
                    .descendant { it.id == footnoteHref }
                    .firstOrNull()
                    .apply { if (this == null) throw Exception("Footnote $footnoteHref not found") }!!
                    .children()
                    .mapIndexed { i, el ->
                        if (i == 0 && el is Anchor) {
                            null
                        } else if (i == 1 && el is Text) {
                            Text(el.text.substring(2))
                        } else {
                            el
                        }
                    }.filterNotNull()
                val footnote = Footnote(footnoteId).apply {
                    footnoteBody.forEach { this.appendChild(it) }
                }
                footNoteSpan.insertAfter(footnote)
                footNoteSpan.remove()
                footnotes[footnoteHref] = footnote
            }
        this.descendant { it is Span && it.roles.contains("footnoteref") }
            .forEach { footnoteRefSpan ->
                val footnoteToRefHref = footnoteRefSpan.descendant { it is Anchor }.map { it as Anchor }
                    .firstOrNull()
                    .apply { if (this == null) throw Exception("Anchor to footnote reference not found") }!!
                    .href
                    .substring(1)
                val footnoteToRef = footnotes[footnoteToRefHref]
                    ?: throw Exception(
                        "Can't find footnote ${footnoteRefSpan.id} to reference"
                    )
                val footnoteRef = FootnoteRef(footnoteToRef)
                footnoteRefSpan.insertAfter(footnoteRef)
                footnoteRefSpan.remove()
            }
    }

    fun Node.wrapBlockImages() {
        this.descendant { image ->
            image is Image &&
                    image.ancestor { it.roles.contains("imageblock") }
                        .isNotEmpty()
        }.forEach { image ->
            val imageBlock = image.ancestor { it.roles.contains("imageblock") }.firstOrNull()
            val p =
                Paragraph().apply { roles("imageblock-paragraph") }
            if (imageBlock?.id != null) {
                p.id = imageBlock.id
                imageBlock.id = "image-block--${imageBlock.id}"
            }
            image.insertAfter(p)
            p.appendChild(image)
        }
    }

    fun Node.transformDefinitionList() {
        this.descendant { it.sourceTagName == "dt" }.forEach { dt ->
            dt.apply {
                val paragraph = Paragraph().apply {
                    dt.children().forEach { appendChild(it) }
                    roles("definition-term")
                }
                insertAfter(paragraph)
                remove()
            }
        }
    }

    fun Node.transformExamples() {
        descendant { it.roles.contains("exampleblock") }.forEach { exampleBlock ->
            exampleBlock.descendant { it.roles.contains("content") }.forEach { exampleContent ->
                exampleContent.appendFirstChild(
                    Paragraph().apply {
                        roles("keep-with-next", "horizontal-line-for-example")
                    }
                )
                exampleContent.p { roles("horizontal-line-for-example") }
                    .previous().roles("keep-with-next")
            }
        }
    }

    override fun Node.normalizeAll() {
        trimSimpleTdContent()
        extractToc()
        wrapTableCellInlineContent()
        wrapTitleInlineContent()
        wrapBlockImages()
        normalizeHeaders()
        transformAdmonitions()
        transformExamples()
        transformDefinitionList()
        extractFootnotes()
        wrapTablesInLists()
        normalizeImageDimensions()
        setImageBestFitDimensions()
    }
}