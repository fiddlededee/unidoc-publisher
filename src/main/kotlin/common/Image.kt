package common

import fodt.iterable
import fodt.parseStringAsXML
import fodt.xpath
import model.LengthUnit
import org.w3c.dom.Attr
import java.util.*
import javax.imageio.ImageIO
import java.io.ByteArrayInputStream
import java.io.File
import javax.imageio.IIOException

data class ImageMeta(val width: Float, val height: Float, val unit: LengthUnit)

fun imageMeta(src: String): ImageMeta {
    val base64Regex = """^data:image/(.*);base64,(.*)$""".toRegex()
    val matchResult = base64Regex.matchEntire(src)
    val type = if (matchResult == null) src.substringAfterLast('.', "") else
        matchResult.groupValues[1]
    return if (type == "svg" || type == "svg+xml") svgImageMeta(matchResult, src) else
        binaryImageMeta(matchResult, src)
}

private fun svgImageMeta(matchResult: MatchResult?, src: String): ImageMeta {
    val svgXml = run {
        if (matchResult == null) {
            File(src).readText()
        } else {
            val base64Image = matchResult.groupValues[2]
            String(Base64.getDecoder().decode(base64Image))
        }
    }
    val viewBox = svgXml.parseStringAsXML().xpath("/svg:svg/@viewBox").iterable().map { it as Attr }
        .firstOrNull()?.value
        ?.split(" ") ?: throw Exception("ViewBox not found in svg file")
    return ImageMeta(viewBox[2].toFloat(), viewBox[3].toFloat(), LengthUnit.parrots)
}

private fun binaryImageMeta(matchResult: MatchResult?, src: String): ImageMeta {
    val img = run {
        if (matchResult == null) {
            try {
                ImageIO.read(File(src))
            }catch (e : IIOException) {
                throw IIOException("Can't read input file ${src}")
            }
        } else {
            val base64Image = matchResult.groupValues[2]
            val imageBytes: ByteArray = Base64.getDecoder().decode(base64Image)
            ImageIO.read(ByteArrayInputStream(imageBytes))
        }
    }
    return ImageMeta(img.width.toFloat(), img.height.toFloat(), LengthUnit.px)
}
