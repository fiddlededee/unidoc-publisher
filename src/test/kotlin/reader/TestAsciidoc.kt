package reader

import common.asciidocAsHtml
import common.htmlMarkupAsHtmlNode
import common.parseAsciidoc
import model.Document
import model.Node
import model.Text
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class TestAsciidoc {
    @Test
    fun testSimpleAsciidoc() {
        """
            = Sample Asciidoc 
           
            == Heading 1
            
            [.unnecessary]
            This paragraph should not appear in result
            
            This paragraph should appear in result
            
            == Heading 2
            
            Some more paragraph
        """
            .trimIndent()
            .asciidocAsHtml()
//            .apply { println(this) }
            .htmlMarkupAsHtmlNode()
            .selectAtXpath("/html/body")!!
            .parseAsciidoc().apply {
                this.descendant { it.roles.contains("unnecessary") }
                    .forEach { it.remove() }
                this.descendant { it.id == "footer" }
                    .forEach { it.remove() }
            }.apply { normalizeWhitespaces() }
            .output()
            .apply { Approvals.verify(this) }
    }
}


fun Document.output(): String {
    val astAsText = StringBuilder()
    fun outputNode(node: Node, level: Int): String {
        if (node is Text) {
            astAsText.appendLine("${" ".repeat(level * 2)}text: [${node.text}]")
        } else {
            astAsText.appendLine(
                "${" ".repeat(level * 2)}${node::class.java.simpleName} id=${node.id}" +
                        " roles=${node.roles.joinToString(",")}"
            )
            node.children().forEach {
                outputNode(it, level + 1)
            }
        }
        return astAsText.toString()
    }
    outputNode(this, 0)
    return astAsText.toString()
}
