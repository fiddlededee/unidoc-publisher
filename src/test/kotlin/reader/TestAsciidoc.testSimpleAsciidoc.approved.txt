Document id=null roles=
  OpenBlock id=header roles=
    Heading id=null roles=
      text: [Sample Asciidoc]
  OpenBlock id=content roles=
    Heading id=_heading_1 roles=
      text: [Heading 1]
    Paragraph id=null roles=
      text: [This paragraph should appear in result]
    Heading id=_heading_2 roles=
      text: [Heading 2]
    Paragraph id=null roles=
      text: [Some more paragraph]
